Enhancement/59 implement typescript

Rennzie opened the pull request:
closes: https://github.com/GeoTIFF/georaster-layer-for-leaflet/issues/59


Opening PR for easier discussions.

Rennzie commented on src/georaster-layer-for-leaflet.ts:
```diff
@@ -0,0 +1,592 @@
+      this.debugLevel = options.debugLevel;
+      if (this.debugLevel >= 1) console.log("georaster:", options);
+
+      //! note: Values is missing in type. What should it be? number[][][] - SFR 2021-01-20
+      if (this.georasters.every((georaster: Georaster) => typeof georaster.values === "object")) {
+        this.rasters = this.georasters.reduce((result: number[], georaster: Georaster) => {
+          result = result.concat(georaster.values);
+          return result;
+        }, []);
+        if (this.debugLevel > 1) console.log("this.rasters:", this.rasters);
+      }
```
@DanielJDufour, This block is confusing me a bit. You mentioned that `georaster.values` could be undefined if the `source` is `url`. The compiler is complaining on line 88 that the concat cannot have an `| undefined` type. Is there a different way we could write this so that we are sure `values` is defined?

DanielJDufour replied:
Would it be possible to do 

`result.concat(georaster.values || []);`

or:
```js
        this.rasters = this.georasters.reduce((result: number[], georaster: Georaster) => {
          if (georaster.values) result = result.concat(georaster.values);
          return result;
        }, []);
```

DanielJDufour replied:
You're correct that values should be `number[][][] `.  It's a three dimensional array of numbers.  The first dimension is band, second is row, and third is column.

DanielJDufour replied:
But it could also be undefined if georaster isn't loading all the values into memory.

Rennzie replied:
Neither of those worked I'm afraid. What is `this.rasters` expected to be? Is it a four dimensional array given the `georaster.values` is a `number[][][]`

Could it be rewritten as: 

```typescript
 this.rasters = this.georasters.map((georaster: Georaster) => georaster.values );
```

Effectively pulling all the values from the georaster into an array.

Rennzie replied:
I tried my solution above and it didn't work until I changed it to: 

```javascript
this.rasters = this.georasters.map((georaster: Georaster) => georaster.values )[0];
```

It's confusing me a bit why that necessary but it seems to work. I'm wary of changing the implementation but would that do what you expected it too?

DanielJDufour replied:
I believe this would return the values for the first georaster only.  For example, with Landsat8 imagery on AWS, each band is its own GeoTIFF file.  In this case, the above code would only pull the first file.

To get around the compiler/signature issues could we default the value to null in the code and update the signature to `null | [][][]`?  Seems like defaulting values to null instead of undefined is a way to get around undefined issues.

Rennzie commented on src/georaster-layer-for-leaflet.ts:
```diff
@@ -0,0 +1,592 @@
+    const yMinOfTileInMapCRS = boundsOfTile.getSouth();
+    const yMaxOfTileInMapCRS = boundsOfTile.getNorth();
+
+    let rasterPixelsAcross, rasterPixelsDown;
+    if (inSimpleCRS || this.projection === EPSG4326) {
+      // width of the Leaflet tile in number of pixels from original raster
+      rasterPixelsAcross = Math.ceil((xMaxOfTileInMapCRS - xMinOfTileInMapCRS) / pixelWidth);
+      rasterPixelsDown = Math.ceil((yMaxOfTileInMapCRS - yMinOfTileInMapCRS) / pixelHeight);
+      // note: why do we check for getProjector() - SFR 2021-01-20
+    } else if (this.getProjector()) {
+      const projector = this.getProjector();
+      // convert extent of Leaflet tile to projection of the georaster
+      const topLeft = projector.inverse({ x: xMinOfTileInMapCRS, y: yMaxOfTileInMapCRS });
+      const topRight = projector.inverse({ x: xMaxOfTileInMapCRS, y: yMaxOfTileInMapCRS });
+      const bottomLeft = projector.inverse({ x: xMinOfTileInMapCRS, y: yMinOfTileInMapCRS });
+      const bottomRight = projector.inverse({ x: xMaxOfTileInMapCRS, y: yMinOfTileInMapCRS });
+
+      rasterPixelsAcross = Math.ceil(Math.max(topRight.x - topLeft.x, bottomRight.x - bottomLeft.x) / pixelWidth);
+      rasterPixelsDown = Math.ceil(Math.max(topLeft.y - bottomLeft.y, topRight.y - bottomRight.y) / pixelHeight);
+    }
+
+    const { resolution } = this.options;
+
+    // prevent sampling more times than number of pixels to display
+    const numberOfSamplesAcross = Math.min(resolution, rasterPixelsAcross);
+    const numberOfSamplesDown = Math.min(resolution, rasterPixelsDown);
```
@DanielJDufour, the compiler complains that `rasterPixelsAcross` & `rasterPixelsDown` could be undefined as they are set within the `if` block (lns: 237-252). `Math.min` expects a type number only. Is it possible to give the `rasterPixels*` variables a default of 0? Or to move lines 257&258 within the `if` block?

DanielJDufour replied:
Yes, you could definitely default rasterPixelsAcross to zero.  Or you could default it to 256 (which is the standard map tile size).  Either way.

Rennzie replied:
Great. And I can do the same for `rasterPixelsDown`?

DanielJDufour replied:
Seems like you solved this, so I'll resolve this thread.

Rennzie commented on src/georaster-layer-for-leaflet.ts:
```diff
@@ -0,0 +1,592 @@
+              }
+
+              let values = null;
+              if (tileRasters) {
```
The compiler is complaining that `tileRasters` is used before being defined. Underwhat conditions is `tileRaster` undefined and what should happen when it is? 

Could we declare a single variable `tileRasters` and assign it either to `this.rasters` or `this.getRasters` depending on some condition, then use that variable in the `for` block instead of checking what exists again on line 324-332? That way it will always be defined.

DanielJDufour replied:
tileRasters is undefined if the georaster is loading from a file that is loaded from an in-memory file.  tileRasters will be defined when we are using a georaster loaded from a url and tileRasters will refer to just the parts of the raster that are fetched to fulfill the tile request.

DanielJDufour replied:
@Rennzie , could we default tileRasters to `null`?  Would that make the compiler happy?

Rennzie replied:
Yeah nice one. Defaulting to `null` solved it.

Rennzie commented on src/georaster-layer-for-leaflet.ts:
```diff
@@ -0,0 +1,592 @@
+    const width = Math.pow(2, z);
+
+    // check one world to the left
+    const leftCoords = L.point(x - width, y);
+    leftCoords.z = z;
+    if (layerBounds.overlaps(this._tileCoordsToBounds(leftCoords))) return true;
+
+    // check one world to the right
+    const rightCoords = L.point(x + width, y);
+    rightCoords.z = z;
```
`L.point` is of type `Coords` which is simply an d `x` and `y`. What is the `z` value for? Is it needed downstream? I can augment the point type for this scenario in this example if need be.

DanielJDufour replied:
I believe the z value is needed here for _tileCoordsToBounds to convert the point to a rectangle representing the tile.  The z represents the zoom level.  X and Y aren't latitude/longitude in this case, but how many mercator tiles from the left and top the tile represented by the coords is.

Rennzie replied:
Cool. Casting L.point to a coord type where Coord simply extends Point by including the z. 

```typescript
const leftCoords = L.point(x - width, y) as Coords;
```

Rennzie commented:
Hey @DanielJDufour, 

Sorry this has taken so long. In the interest of moving forward I've solved the last typing issues by silencing the compiler `@ts-ignore`. The code worked before so this won't stop it working correctly. This gives us a chance to test the repo and see what's needed before a full release.

As it stands there are three things outstanding:

1. Get all the types output to `dist/index.s.ts`. I cannot figure out why this isn't happening but I don't think it's a blocker. 
2. Browserify: I've not set this up as I'm sure Webpack can do the job. I'd like to test this with a pre-release
3. Comment tidy up: I'd like your feedback on the code before I clean the last bits out. 

It would be great if you could have a look at and comment on the PR. If you've got the time to test it, that would be even better. Could I ask that you release to npm as a beta version. I'm writing a blog post on implementing COGs with a colleague and will be able to test it in a Code Sandbox.  I'll also test loading from a `<script/>` tag which, if I have it right, is the use case for browserify?

I hope you're doing well and keeping safe. 

Sean

DanielJDufour commented:
@Rennzie , awesome.  I'm busy through the weekend, but should have some time next week to do a look at it.  Thank you!

Rennzie commented:
Great stuff. Thanks @DanielJDufour

DanielJDufour commented:
I haven't used .npmignore before (from what I can remember).  What files are included when you run `npm pack`?  I just want to make sure we are excluding the test files and other un-needed files from the npm package.

DanielJDufour commented on src/georaster-layer-for-leaflet.ts:
```diff
@@ -308,11 +328,12 @@ const GeoRasterLayer = L.GridLayer.extend({
                 });
               } else {
-                done && done("no rasters are available for, so skipping value generation");
+                done && done(Error("no rasters are available for, so skipping value generation"));
```
cool.  I thought you needed a `new` before Error, but researched it and I was totally wrong. Thanks for showing me this! 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Error#function_call_or_new_construction

Rennzie replied:
Ah, I didn't have that sort of clarity. Thanks for the link.

DanielJDufour commented on src/types/index.d.ts:
```diff
@@ -0,0 +1,108 @@
+/* eslint-disable camelcase */
+import type { GridLayerOptions, Coords, DoneCallback } from "leaflet";
+
+export type PixelValueToColorFn = (values: number[]) => string;
```
I think there's a slight typo here and it should be PixelValue**s**ToColorFn?

Rennzie replied:
Yup, great catch.

DanielJDufour commented on src/types/index.d.ts:
```diff
@@ -0,0 +1,108 @@
+export type PixelValueToColorFn = (values: number[]) => string;
+
+export interface GeorasterLayerOptions extends GridLayerOptions {
+  georasters: Georaster[];
```
Should this be `georasters?:` and `georaster?:` because user could do either one?

Rennzie replied:
I've totally re-typed this. I used the file locally in a react project and learned a few things.

Rennzie replied:
Essentially the type ensures either, `georaster` or `georasters` is defined.

DanielJDufour commented on src/types/index.d.ts:
```diff
@@ -0,0 +1,108 @@
+  georasters: Georaster[];
+  georaster: Georaster;
+  resolution?: number;
+  debugLevel?: 0 | 1;
```
This should be `debugLevel?: 0 | 1 | 2`

DanielJDufour replied:
Sorry I haven't been consistent in the source code and sometimes do if (debugLevel > ...) and sometimes (debugLevel >= ... )

Rennzie replied:
Good to know. No worries.

DanielJDufour commented on src/types/index.d.ts:
```diff
@@ -0,0 +1,108 @@
+  georaster: Georaster;
+  resolution?: number;
+  debugLevel?: 0 | 1;
+  pixelValuesToColorFn?: PixelValueToColorFn;
```
Should be PixelValuesToColorFn here

Rennzie replied:
done

DanielJDufour commented on src/types/index.d.ts:
```diff
@@ -0,0 +1,108 @@
+  retain?: boolean;
+};
+
+export type GeorasterSource = {
```
I think we can remove GeorasterSource because that's used by a georaster dependency (geotiff.js) and not georaster or georaster-layer-for-leaflet.

DanielJDufour commented on src/types/index.d.ts:
```diff
@@ -0,0 +1,108 @@
+  _blob_is_available: boolean;
+  _data: string;
+  _geotiff: Record<string, unknown> | undefined;
+  bigTiff: boolean;
```
I think bigTiff ... source are properties of _geotiff and not the top level object

DanielJDufour commented on tests/arrayBuffer.html:
```diff
@@ -1,8 +1,8 @@
   <head>
-    <meta name="viewport" content="width=device-width, initial-scale=1">
-    <link rel="stylesheet" href="https://unpkg.com/leaflet@1.0.3/dist/leaflet.css"/>
+    <meta name="viewport" content="width=device-width, initial-scale=1" />
```
meta doesn't need the closing `/` 
https://www.w3schools.com/tags/tag_meta.asp

Rennzie replied:
Hmmm, my linter added without me knowing. Do you specifically want it without?

DanielJDufour replied:
naw. it's okay if the linter's automatically adding it. maybe we can look at removing it later (for another PR)

DanielJDufour commented on tests/arrayBuffer.html:
```diff
@@ -1,8 +1,8 @@
-    <meta name="viewport" content="width=device-width, initial-scale=1">
-    <link rel="stylesheet" href="https://unpkg.com/leaflet@1.0.3/dist/leaflet.css"/>
+    <meta name="viewport" content="width=device-width, initial-scale=1" />
+    <link rel="stylesheet" href="https://unpkg.com/leaflet@1.0.3/dist/leaflet.css" />
```
I was wrong to originally include the `/` at the end of the link and it can be removed:
https://www.w3schools.com/tags/tag_link.asp

DanielJDufour commented on tsconfig.json:
```diff
@@ -0,0 +1,73 @@
+
+    /* Basic Options */
+    // "incremental": true,                   /* Enable incremental compilation */
+    "target": "es2020" /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */,
```
I'd like to target older browsers, so I think this should be ES5.

Rennzie replied:
This is what `tsc` put out, we then run it through babel to get the `dist/**bundled.js` version. That bundled version is used by webpack for the `dist/**min.js` version. I'm not sure we need all these outputs but I did it to preserve what was already output. I kept this as `es2020` so a use could copy it directly into their project if need be.

DanielJDufour replied:
that's brilliant!  thank you for the explanation

DanielJDufour commented on tsconfig.json:
```diff
@@ -0,0 +1,73 @@
+    // "incremental": true,                   /* Enable incremental compilation */
+    "target": "es2020" /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */,
+    "module": "es2020" /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'. */,
+    "lib": ["es2017", "es7", "es6", "dom"] /* Specify library files to be included in the compilation. */,
```
This doesn't increase the bundle size, right?

Rennzie replied:
No it shouldn't.

DanielJDufour commented on tsconfig.json:
```diff
@@ -0,0 +1,73 @@
+    // "rootDir": "./",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
+    // "composite": true,                     /* Enable project compilation */
+    // "tsBuildInfoFile": "./",               /* Specify file to store incremental compilation information */
+    "removeComments": false /* Do not emit comments to output. */,
```
what's the argument here for not removing comments?

Rennzie replied:
The `tsc` output is essentially a javascript version of `georaster-layer-for-leaflet.ts`. I thought keeping the comments in would help anyone reading the file. Happy to remove if you don't think it's necessary.

DanielJDufour commented on webpack.config.js:
```diff
@@ -1,23 +1,30 @@
     filename: "georaster-layer-for-leaflet.min.js",
-    path: path.resolve(__dirname)
+    path: path.resolve(__dirname, "dist"),
+    library: "libraryStarter",
```
shouldn't this be `library: GeoRasterLayer`?
https://webpack.js.org/configuration/output/#outputlibrary

Rennzie replied:
Yes definitely. Missed that

DanielJDufour commented on webpack.config.js:
```diff
@@ -1,23 +1,30 @@
+    modules: ["node_modules"]
+  },
+  externals: {
+    leaflet: { root: "L", commonjs: "leaflet", amd: "leaflet", commonjs2: "leaflet" }
```
so does this remove Leaflet from the build but still allows us to do import from leaflet?  Very nifty!

Rennzie replied:
Exactly right. Anything in here needs to be declared as a PeerDependancy. Then it's as use described. Any user of the lib will need to install `leaflet`, but it's not included in our build.

DanielJDufour reviewed (changes requested):
This was a massive PR and a ton of work.  Thank you so much!  Please review my comments and let me know what you think?

I'm also going to give some more thought to browserify and what to do there.  It might be feasible to leave the current browserify build there in case people are depending on it, but just not update/rebuild it anymore.

DanielJDufour commented:
I also learned a bunch about TS and other some cool tricks.  Thank you!

Rennzie commented:
Hey @DanielJDufour , sorry I've not gotten around to this yet. We've had some upheaval at work that needed attention. I'll hopefully have some time this week.

Rennzie commented on package.json:
```diff
@@ -2,18 +2,23 @@
+    "build:main": "tsc",
+    "build:webpack": "webpack --config=webpack.config.js",
+    "build": "npm run build:babel && npm run build:webpack && npm run build:main",
+    "dev": "concurrently \"npm run serve\" ",
```
Ah nice catch. That needs to be set up. The test files are using the webpack build. I'll update to us `webpack --watch` as well as `serve`.

Rennzie commented on package.json:
```diff
@@ -47,16 +52,29 @@
   "peerDependencies": {
-    "georaster": "^1.1.0"
-  }
+    "georaster": "^1.1.0",
```
👍🏻

Rennzie commented on src/georaster-layer-for-leaflet.ts:
```diff
@@ -1,17 +1,28 @@
 import getProjString from "utm-utils/src/getProjString";
+import type { Coords, DoneCallback, LatLngTuple } from "leaflet";
+import type {
+  GeorasterLayerOptions,
```
Yeah sure. I haven't been very consistent for that. Will update them all

Rennzie commented on src/georaster-layer-for-leaflet.ts:
```diff
@@ -70,18 +82,21 @@ const GeoRasterLayer = L.GridLayer.extend({
+        console.log("%c GEORASTERS", "color: red", this.georasters);
+        this.rasters = this.georasters.map((georaster: Georaster) => georaster.values)[0];
+
+        // this.rasters = this.georasters.reduce((result: number[][], georaster: Georaster) => {
```
Ok fair, I still don't fully understand this code bock and hoped I hadn't changed it. I've reverted the code and used a `@ts-ignore`. The `if()` already checks each georaster has defined values.

Rennzie commented on src/georaster-layer-for-leaflet.ts:
```diff
@@ -70,18 +82,21 @@ const GeoRasterLayer = L.GridLayer.extend({
       this.scale = chroma.scale();
 
-      L.setOptions(this, options);
+      L.Util.setOptions(this, options);
```
The  compiler was throwing and error. I'm not sure if it's an incorrect type in the `definitely-typed` repo for `leaflet`. That said, `setOptions` comes from the `Util` class originally so this is sounds.

Rennzie commented on src/georaster-layer-for-leaflet.ts:
```diff
@@ -130,7 +145,8 @@ const GeoRasterLayer = L.GridLayer.extend({
     */
-    const rasterCoordsForTileCoords = (h, w) => {
+    // todo:<ts-conversion> check this return statement is correct - SFR 2021-01-19
+    const rasterCoordsForTileCoords = (h: number, w: number): { x: number; y: number } | undefined => {
```
That comment was old. But, using an `else` and a `null` type is better I think.

Rennzie commented on src/georaster-layer-for-leaflet.ts:
```diff
@@ -254,7 +274,7 @@ const GeoRasterLayer = L.GridLayer.extend({
     // render asynchronously so tiles show up as they finish instead of all at once (which blocks the UI)
     setTimeout(async () => {
-      let tileRasters;
+      let tileRasters: number[][][] | null = null;
```
haha thanks

Rennzie commented on src/georaster-layer-for-leaflet.ts:
```diff
@@ -308,11 +328,12 @@ const GeoRasterLayer = L.GridLayer.extend({
                 // get value from array with data for entire raster
-                values = rasters.map(band => {
+                values = rasters.map((band: Georaster) => {
+                  // @ts-ignore it works but needs validation. Current error: Type 'null' cannot be used as an index type
```
That didn't work unfortunately. This block above it:

```javascript
         let yInRasterPixels: number | null = null;
          if (inSimpleCRS || this.projection === EPSG4326) {
            yInRasterPixels = Math.floor((yMaxOfLayer - lat) / pixelHeight);
          } else {
            yInRasterPixels = null;
          }
```

Is explicitly setting null in the `else`. 

We then drop out the `if` block and there is no check in the `for (let w = 0; w < numberOfSamplesAcross; w++) {` loop testing if `yInRasterPixels` is null. From the compilers point of view it's still possible that it's null and being used to index and array. That's what fails. 

It could be solve by moving that `for` loop into the `if` statement. Or defaulting `yInRasterPixels` to 0. Any ideas?

DanielJDufour replied:
Let's just set yInRasterPixels to 0.  I'm definitely open to optimizing/refactoring the logic of where to put the for loops later, but don't want to postpone merging your PR.

Rennzie commented on src/georaster-layer-for-leaflet.ts:
```diff
@@ -216,11 +234,13 @@ const GeoRasterLayer = L.GridLayer.extend({
       // width of the Leaflet tile in number of pixels from original raster
       rasterPixelsAcross = Math.ceil((xMaxOfTileInMapCRS - xMinOfTileInMapCRS) / pixelWidth);
       rasterPixelsDown = Math.ceil((yMaxOfTileInMapCRS - yMinOfTileInMapCRS) / pixelHeight);
+      // note: why do we check for getProjector() - SFR 2021-01-20
```
I can't remember what triggered that comment. But I've change it to simple `else`.

Rennzie commented on src/georaster-layer-for-leaflet.ts:
```diff
@@ -164,38 +180,40 @@ const GeoRasterLayer = L.GridLayer.extend({
 
     const getValuesOptions = {
-      bottom: bottomRight.y,
+      bottom: bottomRight?.y,
```
The problem lies with the existing `else` where `rasterCoordsForTileCoords` on line 147 could return `null`. If there are better defaults it could return, then the optional chaining could be removed.

Rennzie commented on src/georaster-layer-for-leaflet.ts:
```diff
@@ -1,17 +1,28 @@
-/* global L, proj4 */
+/* global proj4 */
 import "regenerator-runtime/runtime";
+import * as L from "leaflet";
```
Yeah agreed. Webpack take care of that for us and `leaflet` is a PeerDependancy, so users will have to include their own versions.

Rennzie commented:
Hey @DanielJDufour , I'm sorry this has taken me so long. Hopefully it's in a place now thats workable. I've addressed all your comments, either with fixes in the code or further thoughts for discussion. Would love any further feedback you've got. The detailed reviews have been amazing to work with.

DanielJDufour commented on src/types/index.d.ts:
```diff
@@ -1,16 +1,24 @@
+  pixelValuesToColorFn?: PixelValuesToColorFn;
 }
 
+// Ensures at least one of the georaster[s] options is defined while being ok the other is not
```
Wow this is awesome.

Rennzie replied:
Super cool right. I still need to look at it a few times to make sure I understand what it's doing. But its very powerful I find.

DanielJDufour reviewed (changes requested):
Hi, @Rennzie .  I responded to your comments.  I think we just need to set the yInRasterPixels to 0 and then we are good to merge into master!

After we merge into master, there's a few minor things that I'll do on my end.  And we will want to thoroughly test it before publishing to NPM.

Do you know if there is a way to publish a release candidate to NPM, but make it so that when people run `npm install georaster-layer-for-leaflet` it's the last stable version and make sure `unpkg` uses the latest stable version, too?  I haven't published rc or beta versions to NPM before, so the process is a little new to me.  Alternatively, we could publish a 1.0.0-rc version to NPM and just test intensively afterwards in case we broke anything.

Rennzie commented:
Hey @DanielJDufour thanks for reviewing so quickly. I'll set `yInRasterPixels` to 0 which will resolve the type error. 

I'm new to beta releases too, but you can do `npm publish --tag [alpha, beta, rc, next]` which tells npm it's not the `latest`. Then  `npm install georaster-layer-for-leaflet` will install the `latest` and not the tagged version. You'd have to explicitly do `npm install georaster-layer-for-leaflet@alpha` to get the tagged version. Reading the unpkg docs it respects this behaviour too.

Rennzie commented:
Ok, all done and pushed.

DanielJDufour commented:
Awesome.  Would you be able to remove the "draft" from this PR, so I can merge it into master.  Also, Github is showing that there is a merge conflict.  Would you be able to resolve and then I can merge?

<img width="1126" alt="Screen Shot 2021-03-21 at 12 53 03 PM" src="https://user-images.githubusercontent.com/4313463/111913416-61506500-8a44-11eb-824f-6b9eca565705.png">

Rennzie commented:
@DanielJDufour conflicts all fixed and draft status removed.

DanielJDufour commented:
@Rennzie , just merged.  I just have a few small things to do before republishing.  Thank you for all your awesome work!
